### DISCUSSION TITLE
npm binary missing on Renovate v39.10.0

### DISCUSSION THREAD

**Discussion by juancarlosjr97** [renovatebot/renovate#32462] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

39.10.10

### Please tell us more about your question or problem

Running Renovate from Docker, we received an update from our Renovate to migrate from 38.142.7 to 39.10.0. The issue occurs on v39.10.0, and npm is no longer part of the Renovate build.

### Logs (if relevant)

<details><summary>Logs renovate:38.142.7</summary>

```
~ docker run -it --entrypoint="" -u root docker.io/renovate/renovate:38.142.7 sh
sh: 3: Syntax error: "(" unexpected
# node -v
v20.18.0
# npm -v
10.8.2
# which npm
/usr/local/bin/npm
```

</details>

<details><summary>Logs renovate:39.10.0</summary>

```
docker run -it --entrypoint="" -u root docker.io/renovate/renovate:39.10.0 sh
sh: 3: Syntax error: "(" unexpected
# node -v
v22.11.0
# npm -v
sh: 2: npm: not found
# which npm
# 
```

</details>

**Comment by rarkins** [comment, marked as answer]
This is intended. Use the -full image if you need package managers preinstalled.

**Reply by juancarlosjr97** [reply]
Thanks for clarifying.

So, is this the breaking change? From the release notes [here](https://github.com/renovatebot/renovate/releases/tag/39.0.0), I could only see a reference to the full image, not the slim image, or when this change happened.

> Dockerfile: Renovate -full images will now use binarySource=global. Set RENOVATE_BINARY_SOURCE=install to revert to dynamic tool installation.

**Reply by rarkins** [reply]
The installed package managers in the slim image were never part of Renovate's API or docs. It's only breaking if it's a change to documented functionality or capabilities

**Reply by juancarlosjr97** [reply]
I see... So, it is just that npm was included unintentionally in the slim image before v39, and now it is no longer there.

**Reply by rarkins** [reply]
Yes, because renovate itself runs on node

**Reply by juancarlosjr97** [reply]
Thanks, @rarkins. interesting change! :)

It is still unclear where npm was coming from before v39.x. But diving deeper does not feel like a good use of time, as it seems that npm’s inclusion was an unintentional bonus. :)

**Reply by Taucher2003** [reply]
I had the same problem. And because I just need `npm`, I didn't want to switch to the full image which contains a lot more than I need. But there is a `install-npm` script available, so I just added `install-npm node $(node --version)` as a step in our pipeline which installs the version of npm that matches the node version of Renovate and it works great again. (and I do realize that this is also not documented and might break in the future, but it prevents pulling an image that is a lot larger than what I need)

**Reply by astellingwerf** [reply]
It would have been great if this breaking change was documented.

**Reply by viceice** [reply]
node is again available in path

**Reply by astellingwerf** [reply]
`node` is, but `npm` and `npx` are not. This change silently broke our post upgrade tasks.

**Reply by viceice** [reply]
yes, they're not included in the slim image for already said reason. you can use the full image which has almost all tools installed
